workers/ohsome_quality_analyst/indicators/mapping_saturation/fit.py

```diff
@@ -1,111 +1,59 @@
-"""Fit sigmoid functions to data.
+"""Run all models and choose best fit based on Mean Absolute Error."""
 
-Nonlinear least squares optimization is used to fit data to sigmoid functions which
-include up to four jumps.
 
-Best fitting function is chosen by the applying mean-squared error criterion.
-"""
-
-import logging
 from dataclasses import dataclass
-from typing import Tuple
+from typing import List
 
-import numpy as np
-from scipy.optimize import curve_fit
+from numpy import float64, ndarray
 
 from ohsome_quality_analyst.indicators.mapping_saturation import metrics, models
 
 
 @dataclass
 class Fit:
-    model_name: str
-    ydata: np.ndarray
+    asymptote: float64
+    coefficients: dict
+    function_formula: str
+    metric: float64
     metric_name: str
-    metric: np.float64
-    asymptote: np.float64
-
+    model_name: str
+    ydata: ndarray
 
-def get_best_fit(xdata: np.ndarray, ydata: np.ndarray) -> Fit:
-    """Get best fit function based on Mean Squared Error.
 
-    Fit sigmoid_1 to sigmoid_4 to given data and return best fit based on Mean Squared
-    Error.
-    """
-    best_fit = None
+def run_all_models(xdata: ndarray, ydata: ndarray) -> List[Fit]:
     # TODO: Run for every model
-    model = models.Sigmoid()
-    p0 = model.get_initial_guess(xdata, ydata)
-    bounds = model.get_bounds(xdata, ydata)
-    try:
-        # curve_fit: Use non-linear least squares to fit a function, f, to data.
-        # popt: Optimal values for the parameters as array
-        # pcov: The estimated covariance of popt as 2-D array
-        popt, pcov = curve_fit(
-            model.function, xdata=xdata, ydata=ydata, p0=p0, bounds=bounds
+    fits = []
+    for model in (
+        models.Sigmoid(),
+        models.SSlogis(),
+        models.SSdoubleS(),
+        models.SSfpl(),
+        models.SSasymp(),
+        # models.SSmicemen(),
+    ):
+        coef = model.fit(xdata, ydata)
+        ydata_fitted = model.function(xdata, **coef)
+        metric = metrics.mae(ydata, ydata_fitted)
+        fit = Fit(
```

Comment:
I'm a bit confused by these lines. Maybe you can help me and explain what you are doing here. I'm not sure what you "fit" here as you already have `ydata_fitted` a few lines above. 

From my understanding you did the curve fitting already, but here you try to estimate the model performance of your fitted values.

Reply:
Regarding the metrics: Have a closer look at the `metrics.py` file. In this file multiple error metrics are implemented. Its just that the diff is not showing the difference since this PR is not directed at the main branch.

Reply:
I see the reason of the confusion here. The variable names are not very precise.
I renamed the dataclass `Fit` to `FittedModel`. To propose of the dataclass is to collect all relevant data of the model result. So there is no fitting going on in this line, just to collection of the result (fitted values, metric, etc.).

In fact this file (`fit.py`) could probably be refactored into either `models.py` or `indicator.py` (or into both). E.g. all attributes of the `FittedModel` dataclass could actually be attributes of the model class itself.